File: java/memo/GridTraveler.java
import java.util.HashMap;
import java.util.Map;

public class GridTraveler {
    public int gridTravelerRecursive(int m, int n) {
        if (m == 1 && n == 1) return 1;
        if (m == 0 || n == 0) return 0;

        return gridTravelerRecursive(m - 1, n) + gridTravelerRecursive(m, n - 1);
    }

    public void runRecursiveExamples() {
        System.out.println("recursive examples ----------");
        System.out.println(gridTravelerRecursive(5, 5));

        // this is really slow
        // System.out.println(gridTravelerRecursive(18, 18));
    }

    public int gridTravelerMemo(int m, int n, Map<String, Integer> memo) {
        String key = m <= n ? m + "," + n : n + "," + m;

        if (memo.containsKey(key)) return memo.get(key);
        if (m == 1 && n == 1) return 1;
        if (m == 0 || n == 0) return 0;

        // store in memo
        memo.put(key, gridTravelerMemo(m-1, n, memo) + gridTravelerMemo(m, n-1, memo));

        return memo.get(key);
    }

    public void runMemoExamples() {
        System.out.println("memo examples -----------");
        System.out.println(gridTravelerMemo(10,10, new HashMap<String, Integer>()));
    }

    public static void main(String[] args) {
        GridTraveler g = new GridTraveler();
        g.runRecursiveExamples();    
        g.runMemoExamples();
    }
}

File: javascript/memo/grid-traveler.js
/**
 * Say that you are a traveler on a 2D grid. You begin in the top-left corner
 * and your gaol is to travel to the bottom-riht corner. You may only moe down or 
 * right
 * 
 */


 /**
  * Recursive O(2^(n+m)) time complexity solution and O(n+m) space
  * @param {int} m 
  * @param {int} n 
  */
const gridTraveler = (m, n) => {
    if (m === 1 && n === 1) return 1;
    if (m === 0 || n === 0) return 0;

    return gridTraveler(m- 1, n) + gridTraveler(m, n - 1);
};


/**
 * 
 * @param {int} m 
 * @param {int} n 
 * @param {object} memo 
 */
const gridTravelerMemo = (m, n, memo={}) => {
    // flip the key as 1,2 == 2,1 in terms of how many ways to travel through same
    // sub-problem
    const key = m <= n ? m + ',' + n : n + ',' + m;

    // base cases -------------------------
    if (key in memo) return memo[key];
    if (m === 1 && n === 1) return 1;
    if (m === 0 || n === 0) return 0;
    // ------------------------------------

    // store in memo
    memo[key] = gridTravelerMemo(m- 1, n, memo) + gridTravelerMemo(m, n - 1, memo);
 
    // return from memo object
    return memo[key];
};


console.log(gridTravelerMemo(18,18));
console.log(gridTravelerMemo(180,180));
